### PR TITLE
cmd/snap-seccomp: add riscv64 support

### PR DESCRIPTION
**Pull request by xnox**
cmd/snap-seccomp: add riscv64 support

### CODE REVIEW

File: cmd/snap-seccomp/main.go
Comment (mvo5) [inline cmd/snap-seccomp/main.go:192]: Unfortunately this change will break building on trusty. This is something we still want to support if at all possible. Is there a way to avoid this? does the forked repo need an update or something?

Comment (xnox) [reply cmd/snap-seccomp/main.go:192]: @mvo5 this is intentional as explained in the commit message https://github.com/snapcore/snapd/pull/10808/commits/7bd6d7451e529a929c4b2836fc815069ea74ebdd

The last attempt to build snapd on trusty was 2.43.3~14.04 and it never got released. Thus snapd is no longer built on trusty at all. The snapd snap built on xenial, may still be used on trusty.

Comment (mvo5) [reply cmd/snap-seccomp/main.go:192]: Fwiw, I agree with Ian here, if we have some sort of catastrophic security issue and need SRU/ESM snapd back to trusty I would hate to be in a situation where we can't do it. It sounds like we need to involve the security here, either we can drop snapd in trusty from ESM (which is problematic as live-patch on trusty requires snapd) or we need to find a way to either keep working with a patched version of seccomp-golang or get an update of libseccomp 2.2 into trusty.

Comment (mvo5) [reply cmd/snap-seccomp/main.go:192]: Let's explore this a bit more, I created a https://github.com/mvo5/libseccomp-golang/tree/move-to-seccomp-golang-20210920 branch that moves to tip of libseccomp-golang and applies our patches on top, I started a spread run to see if 14.04 still builds with that.

Comment (mvo5) [reply cmd/snap-seccomp/main.go:192]: Fwiw, if we keep using the fork we should move it to "snapcore" out of my personal repo for sure.

Comment (xnox) [reply cmd/snap-seccomp/main.go:192]: @anonymouse64 @mvo5 ESM policies prevent from landing new snapd releases. If you need to build and release new snapd to trusty it would have to be a patch on top of 2.38~14.04

Also, we must continue to test xenial built snap on trusty host. Is this being done?

Comment (xnox) [reply cmd/snap-seccomp/main.go:192]: @anonymouse64 @mvo5 trusty SRU is closed; only ESM security remains open for Trusty, which prohibits accepting anything but High+ targeted security fixes only.

Comment (xnox) [reply cmd/snap-seccomp/main.go:192]: I disagree that upstream libseecomp is used for distro package builds, and not used for re-execed snapd on those hosts.

Surely, we should be patching in to use mvo5's fork on trusty only these days? which is only one combination of them all, instead of forcing to maintain & use a fork on all builds.

Comment (anonymouse64) [reply cmd/snap-seccomp/main.go:192]: @xnox when I said "SRU" I meant SRU via ESM for trusty. The page specifically about ESM on trusty specifically says of snapd under "exclusions":
```
Various packages related to snapd and working with snap packages (not included on installation media). snapd itself is supported to the extent of ensuring livepatch and cvescan remain functional with CVE fixes only for the code representing the operational aspects of livepatch and cvescan functionality.
```

Though actually reading it, it says cvescan too, which is not a classic snap which is a bit troubling since we ignored other security vulnerabilities for trusty since we thought trusty only mattered for livepatch only which is a classic snap so security is kinda nil there. Hmmmmm

Also @xnox we have historically been able to get new version numbers of snapd SRU'd via ESM to xenial, so I think the real answer to this discussion is to discuss options with the security team.

And I totally agree that the best case scenario would be to use mvo's patch only where we need to on trusty and use upstream everywhere else but this is more difficult to maintain for us.

Comment (xnox) [reply cmd/snap-seccomp/main.go:192]: > And I totally agree that the best case scenario would be to use mvo's patch only where we need to on trusty and use upstream everywhere else but this is more difficult to maintain for us.

huh, this is already maintained in a complicated way in current snapd code, where upstream seccomp is used in fedora/debian builds, but not in any others. With a dynamic monkey patch. My proposal is to reverse the monkey patch and move it to trusty-only build path.

**Comment by anonymouse64** [comment]
marking blocked until we agree on the right way forward for maintaining/building with the dependency, we might need to discuss with the security team if we can't figure out how to make this branch happy for all parties without also needing to SRU/ESM things to trusty

**Comment by xnox** [comment]
> marking blocked until we agree on the right way forward for maintaining/building with the dependency, we might need to discuss with the security team if we can't figure out how to make this branch happy for all parties without also needing to SRU/ESM things to trusty

nobody has been doing any ESM releases of snapd to trusty since April 2019. So if you expect that recent snapd releases were making their way to trusty, there is a massive disconnect....

**Comment by anonymouse64** [comment]
> nobody has been doing any ESM releases of snapd to trusty since April 2019. So if you expect that recent snapd releases were making their way to trusty, there is a massive disconnect....

we were not making that assumption to be clear, we haven't been doing SRU's to trusty via ESM because no one has told us about any problems with the livepatch snap on trusty so we haven't had any reason to assume things are broken

**Comment by xnox** [comment]
> > nobody has been doing any ESM releases of snapd to trusty since April 2019. So if you expect that recent snapd releases were making their way to trusty, there is a massive disconnect....
> 
> we were not making that assumption to be clear, we haven't been doing SRU's to trusty via ESM because no one has told us about any problems with the livepatch snap on trusty so we haven't had any reason to assume things are broken

sure. but by ESM policy you will not be allowed to upgrade snapd deb from 2.38~14.04 =>  2.51.1~14.04 anymore. So i'm strugling to comprehend desire to keep master build able on trusty; instead of testing that it is runtime compatible on trusty when built statically linked on xenial.

**Comment by anonymouse64** [comment]
> sure. but by ESM policy you will not be allowed to upgrade snapd deb from 2.38~14.04 => 2.51.1~14.04 anymore. So i'm strugling to comprehend desire to keep master build able on trusty; instead of testing that it is runtime compatible on trusty when built statically linked on xenial.

Well the security team will have to evaluate, rules are made to be broken as I'm sure you know :smile:. It sounds like the next step for this is to discuss with the security team because actually installing the cvescan snap on a fresh trusty VM is "broken" too, to our knowledge we were not responsible for making sure that the cvescan snap works on trusty out of the box, so that may be justification for bumping snapd version to make installing the cvescan snap work out of the box.

**Comment by xnox** [comment]
> > sure. but by ESM policy you will not be allowed to upgrade snapd deb from 2.38~14.04 => 2.51.1~14.04 anymore. So i'm strugling to comprehend desire to keep master build able on trusty; instead of testing that it is runtime compatible on trusty when built statically linked on xenial.
> 
> Well the security team will have to evaluate, rules are made to be broken as I'm sure you know . It sounds like the next step for this is to discuss with the security team because actually installing the cvescan snap on a fresh trusty VM is "broken" too, to our knowledge we were not responsible for making sure that the cvescan snap works on trusty out of the box, so that may be justification for bumping snapd version to make installing the cvescan snap work out of the box.

why is it broken? we do not support fresh trusty installs. And they install new core/snapd snap and re-exec fresh snapd from there, which does work with cvescan, no?

meaning, there is no need to update trusty snapd deb, unless it can't re-exec into new snapd from core/snapd snaps, no?

**Comment by anonymouse64** [comment]
@xnox the issue I have found is that if you launch a trusty system and you don't have any snaps on it, and try to install snapd and then install the cvescan snap, it is installed in a broken state because that old of a version of snapd does not know about base snaps very well, so all you get installed are the core18 and the cvescan snap, without the core snap or the snapd snap. So the snapd version there knows how to do re-exec, it just doesn't know that it needs to install the snapd or core snap when it goes to install the core18 snap because it has a bug assuming that all system interface slots will come from the base snap, which is not an assumption that we ended up carrying through on (not sure if that was ever the intention, but it is how the bug ends up manifesting itself). We face this same issue with Debian 9(?) and Raspbian all the time because they have very old snapd's in their archives too which doesn't know enough about base snaps to hit this same bug.

> we do not support fresh trusty installs

I don't know what this statement means?

**Comment by xnox** [comment]
> > we do not support fresh trusty installs
> 
> I don't know what this statement means?

We don't produce ESM media / images, the intention with ESM that it is an existing system, that already have snaps installed, rather than brand new deployments without nothing on. But obviously users do anything they can.

**Comment by xnox** [comment]
> @xnox the issue I have found is that if you launch a trusty system and you don't have any snaps on it, and try to install snapd and then install the cvescan snap, it is installed in a broken state

that doesn't sound good.

**Comment by anonymouse64** [comment]
> We don't produce ESM media / images, the intention with ESM that it is an existing system, that already have snaps installed, rather than brand new deployments without nothing on. But obviously users do anything they can.

ah in that case, I guess it also depends on how much the security team (who maintains cvescan AFAICT) cares about the different support models for cvescan, I'm not sure if they support someone with ESM activated on their trusty server who one day decides out of the blue to install cvescan and use it, or if they only support cvescan if it came with the image originally when they installed trusty. To be honest, I would be surprised if they _didn't_ support folks just installing it out of the blue, but maybe they have opinions on this.

Regardless, I think the next step for this is indeed to discuss with the security team. I will reach out to them to seek clarification on this after talking with others on the snapd team tomorrow so we can all get on the same page.

File: cmd/snap-seccomp/main.go
Comment (mvo5) [reply cmd/snap-seccomp/main.go:192]: I like the idea of using the fork only on trusty fwiw, I hope now that we have go.mod with some "replace" trickery in there this because actually not too messy. In any case, the update fork branch seems to build so that is (now) also an option and regardless we should switch to it.

**Comment by xnox** [comment]
Some builds are still failing:

```
# github.com/snapcore/snapd/cmd/snap-seccomp
src/github.com/snapcore/snapd/cmd/snap-seccomp/main.go:458:10: undefined: seccomp.ArchRISCV64
```

File: cmd/snap-seccomp/old_seccomp.go
Comment (anonymouse64) [inline cmd/snap-seccomp/old_seccomp.go:30]:
```suggestion
// master is deemed to never be required again. 
```

File: cmd/snap-seccomp/old_seccomp.go
Comment (anonymouse64) [inline cmd/snap-seccomp/old_seccomp.go:2]: hmm this is an interesting way to solve the problem, I think our original idea was to have the packaging scripts for trusty instead add a new line to go.mod like this:

```
replace github.com/seccomp/libseccomp-golang => github.com/mvo5/libseccomp-golang
```

I'm curious if you looked at this at all? What you have is probably fine, but is slightly more opaque than a packaging script modifying go.mod here I think.

**Review by anonymouse64 — COMMENTED** [review]
I think this is probably okay, one question though if you tried the replace directive at all when building on trusty instead of the never used build tag for the old_seccomp.go file

File: cmd/snap-seccomp/old_seccomp.go
Comment (xnox) [reply cmd/snap-seccomp/old_seccomp.go:2]: I couldn't make `go.mod` to be universal for both upstream and trusty at the same time. I.e. `go mod tidy` would continiously drop either mvo5 or the seccomp version of the library. Leaving a gofile with an unused build tag was the only way i figured out to keep `go mod tidy` with both editions of the library.

Comment (anonymouse64) [reply cmd/snap-seccomp/old_seccomp.go:2]: I didn't mean to have a universal go.mod, I just meant to modify the go.mod for trusty alone when building the deb for trusty with like `echo "replace ... " >> go.mod` in the `dh_override_whatever`

Comment (xnox) [reply cmd/snap-seccomp/old_seccomp.go:2]: oh, interesting. I can try hacking that.

Comment (xnox) [reply cmd/snap-seccomp/old_seccomp.go:2]: my concern was to ensure that both copies of vendored lib are available whenever one does snapd-vendor style packaging.

Comment (anonymouse64) [reply cmd/snap-seccomp/old_seccomp.go:2]: hmm, not sure that is super relevant since we mainly just do that for other distros where they don't need the fork of seccomp as they have it packaged already

Comment (xnox) [reply cmd/snap-seccomp/old_seccomp.go:2]: it's exactly how we upload ubuntu-14.04 though, no?

Comment (anonymouse64) [reply cmd/snap-seccomp/old_seccomp.go:2]: uhh I'm confused now, what do you mean about uploading ubuntu-14.04 with vendor sources?

Comment (xnox) [reply cmd/snap-seccomp/old_seccomp.go:2]:
```
pull-lp-source snapd trusty
ls snapd-2.38~14.04/vendor/github.com/mvo5/libseccomp-golang/
```

Something, somewhere, when building 3.0 (native) trusty package downloads all sources when creating source package. Then it is uploaded to launchpad to build. During the build in launchpad it has no network access to github.com to download anything. Meaning to avoid breaking/changing how source package preparation is done, i should ensure that go.mod references mvo5 fork, and it ends up in the source tarballs.

I am also discovering that whilst patches are applied for debian-sid source package builds, they are not applied for the native builds. Thus I will also be trying out to use the `replace` directive during build. Cause i think it is easier than patching; and it is a simple one liner.

**Comment by xnox** [comment]
@mvo5 @anonymouse64 please re-review; trusty support is preserved; libseccomp bindings upgraded everywhere else but trusty; riscv64 support added; and mostly all tests pass everywhere.

File: packaging/ubuntu-14.04/rules
Comment (anonymouse64) [inline packaging/ubuntu-14.04/rules:124]: very nice and straight forward, thanks!

**Review by anonymouse64 — APPROVED** [review]
lgtm, thanks for this

**Review by mvo5 — APPROVED** [review]
Thank you

**Comment by mvo5** [comment]
I think we can remove the blocked flag now, yes?

**Comment by xnox** [comment]
> I think we can remove the blocked flag now, yes?

yeap, just need to get all the tests to pass in all combinations cleanly now.

**Review by anonymouse64 — APPROVED** [review]
still LGTM

File: cmd/snap-seccomp/old_seccomp.go
Comment (bboozzoo) [reply cmd/snap-seccomp/old_seccomp.go:2]: > echo "replace ... " >> go.mod

FWIW this should probably be:
```
go mod edit -replace=github.com/seccomp/libseccomp-golang=github.com/mvo5/libseccomp-golang
```

**Review by bboozzoo — APPROVED** [review]
LGTM

**Comment by xnox** [comment]
google:ubuntu-21.10-64:tests/main/refresh:parallel_strict_fake failure on 21.10-64 looks unrelated.

I think this can be merged now @mvo5

**Comment by xnox** [comment]
Merging master again.

**Pull request merged**
